libyul/ASTLabelRegistry.h

```diff
@@ -0,0 +1,105 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <limits>
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace solidity::yul
+{
+
+/// Instances of the `ASTLabelRegistry` are immutable containers describing a labelling of nodes inside the AST.
+/// Each element of the AST that possesses a label has a `ASTLabelRegistry::LabelID`, with which the label can
+/// be queried in O(1).
+/// Preferred way of creating instances is via `ASTLabelRegistryBuilder` when parsing/importing and
+/// via `LabelIDDispenser` during/after optimization.
+/// Note: There is a special case of ghost ids, which are added during CFG construction.
+/// They do not directly correspond to elements of an AST but live inside the CFG. They are assigned the
+/// special `std::numeric_limits<LabelID>::max` label index and are labelled as `GHOST[{id}]`,
+/// corresponding to their label `id`.
+class ASTLabelRegistry
+{
+public:
+	/// unsafe to use from a different registry instance, it is up to the user to safeguard against this
+	using LabelID = size_t;
+
+	ASTLabelRegistry();
+	ASTLabelRegistry(std::vector<std::string> _labels, std::vector<size_t> _idToLabelMapping);
+
+	std::string_view operator[](LabelID _id) const;
+
+	static bool constexpr empty(LabelID const _id) { return _id == emptyID(); }
+	static LabelID constexpr emptyID() { return 0; }
+	static LabelID constexpr ghostID() { return std::numeric_limits<LabelID>::max(); }
```

Comment:
Isn't this a label index rather than a label ID?
```suggestion
	static size_t constexpr ghostLabelIndex() { return std::numeric_limits<size_t>::max(); }
```

Reply:
Yep, true!